tests/std/tests/GH_002488_promise_not_default_constructible_types/test.cpp

```diff
@@ -0,0 +1,184 @@
+    no_default_or_assign() = delete;
+    explicit no_default_or_assign(int n) : x(n) {
+        ++no_default_or_assign_objects;
+    }
+    no_default_or_assign(const no_default_or_assign& v) : x(v.x) {
+        ++no_default_or_assign_objects;
+    }
+
+    void operator=(const no_default_or_assign&) = delete;
+
+    ~no_default_or_assign() {
+        --no_default_or_assign_objects;
+    }
+
+    int x;
+};
+
+template <class F>
+void assert_throws_future_error(F f, std::error_code errc) {
+    try {
+        f();
+    } catch (const std::future_error& e) {
+        assert(e.code() == errc);
+        return;
+    } catch (...) {
+    }
+    assert(false);
+}
+
+template <class T>
+void run_tests() {
+    using promise = std::promise<T>;
+    using future  = std::future<T>;
+
+    {
+        promise p;
+        p.set_value(T(4));
+        assert(p.get_future().get().x == 4);
+    }
+
+    {
+        promise p;
+        future f = p.get_future();
+        T v(10);
+        p.set_value(v);
+        assert(f.get().x == 10);
+        assert_throws_future_error([&] { p.set_value(v); }, std::future_errc::promise_already_satisfied);
+        assert_throws_future_error([&] { f.get(); }, std::future_errc::no_state);
+        assert_throws_future_error([&] { p.get_future().get(); }, std::future_errc::future_already_retrieved);
+    }
+
+    {
+        promise p;
+        future f = p.get_future();
+        p.set_exception(std::make_exception_ptr(5));
+        try {
+            f.get();
+            assert(false);
+        } catch (int i) {
+            assert(i == 5);
+        } catch (...) {
+            assert(false);
+        }
+    }
+
+    {
+        promise p;
+        future f = p.get_future();
+        p.set_exception(std::make_exception_ptr(3));
+        assert_throws_future_error([&] { p.set_value(T(2)); }, std::future_errc::promise_already_satisfied);
+        try {
+            f.get();
+            assert(false);
+        } catch (int i) {
+            assert(i == 3);
+        } catch (...) {
+            assert(false);
+        }
+    }
+
+    {
+        promise p;
+        future f = p.get_future();
+        std::atomic<int> failures{0};
+        int succeeded    = -1;
+        auto make_thread = [&](int n) {
+            return std::thread([&, n] {
+                try {
+                    p.set_value(T(n));
+                } catch (std::future_error) {
+                    ++failures;
+                    return;
+                }
+                succeeded = n;
+            });
+        };
+        std::thread threads[]{make_thread(0), make_thread(1), make_thread(2), make_thread(3), make_thread(4),
+            make_thread(5), make_thread(6), make_thread(7)};
+
+        for (auto& t : threads) {
+            t.join();
+        }
+
+        assert(failures == 7);
+        assert(succeeded != -1 && f.get().x == succeeded);
+    }
+
+    {
+        (void) std::async(std::launch::async, []() -> T { return T(16); });

```

Comment:
Suggestion: I believe that all of these lambdas written with `[]() -> T` could simply say `[]` as the `T` return type will be deduced, even in C++14 mode. However, if you like the explicit return type, feel free to keep it, it isn't harmful.